lib/OpenQA/Task/Git/Clone.pm

```diff
@@ -0,0 +1,129 @@
+# Copyright SUSE LLC
+# SPDX-License-Identifier: GPL-2.0-or-later
+
+package OpenQA::Task::Git::Clone;
+use Mojo::Base 'Mojolicious::Plugin', -signatures;
+
+use OpenQA::Utils qw(run_cmd_with_log_return_error);
+use Cwd 'abs_path';
+
+sub register ($self, $app, @) {
+    $app->minion->add_task(git_clone => \&_git_clone);
+}
+
+sub _git_clone ($job, $clones) {
+    # $clones is a hashref with paths as keys and urls to git repos as values.
+    # The urls may also refer to a branch via the url fragment.
+   # If no branch is set, the default branch of the remote (if target path doesn't exist yet) or local checkout is used.
+# If the target path already exists, an update of the specified branch is performed, if the url matches the origin remote
+
+    my $app = $job->app;
+    my $job_id = $job->id;
+
+    # Prevent multiple git clone tasks for the same path to run in parallel
+    my @guards;
+    for my $path (sort keys %$clones) {
+        while (-l $path) { $path = abs_path($path); }    # guard symlink target
+        my $guard = $app->minion->guard("limit_git_clone_${path}_task", 7200);
+        return $job->retry({delay => 30 + int(rand(10))}) unless $guard;
+        push(@guards, $guard);
+    }
+
+    my $log = $app->log;
+    my $ctx = $log->context("[#$job_id]");
+
+    # iterate clones sorted by path length to ensure that a needle dir is always cloned after the corresponding casedir
+    for my $path (sort { length($a) > length($b) } keys %$clones) {
+        my $url = $clones->{$path};
+        die("Aber echt ned!") if $path =~ /\.\./;
+        _do_git_clone_path($job, $ctx, $path, $url);
+    }
+}
+
+sub _get_current_branch ($path) {
+    my $r = run_cmd_with_log_return_error(['git', '-C', $path, 'branch', '--show-current']);
+    die("Error detecting current branch for $path") unless $r->{status};
+    my $branch = $r->{stdout};
+    $branch =~ s/\s*$//;    # trim trailing spaces (e.g. newline)
+    return $branch;
+}
+
+sub _get_remote_default_branch ($url) {
+    my $r = run_cmd_with_log_return_error(
+        ['env', 'GIT_SSH_COMMAND="ssh -oBatchMode=yes"', 'git', 'ls-remote', '--symref', $url, 'HEAD']);
+    die("Error detecting remote default branch name for $url")
+      unless $r->{status} && $r->{stdout} =~ /refs\/heads\/(\S+)\s+HEAD/;
+    return $1;
+}
+
+sub _git_clone_url_to_path ($url, $path) {
+    my $r
+      = run_cmd_with_log_return_error(['env', 'GIT_SSH_COMMAND="ssh -oBatchMode=yes"', 'git', 'clone', $url, $path]);
+    die("Failed to clone $url into $path") unless $r->{status};
+}
+
+sub _git_get_origin_url ($path) {
+    my $r = run_cmd_with_log_return_error(['git', '-C', $path, 'remote', 'get-url', 'origin']);
+    die("Failed to get origin url for $path") unless $r->{status};
+    my $url = $r->{stdout};
+    $url =~ s/\s*$//;    # trim trailing spaces (e.g. newline)
+    return $url;
```

Comment:
The trimming should probably its own function so we don't have to repeat this and can just return the trimmed URL in one line. (Maybe you can also use https://docs.mojolicious.org/Mojo/Util#trim or https://perldoc.perl.org/functions/chomp.)